types/meteor/mongo.d.ts

```diff
@@ -128,44 +128,46 @@ declare module "meteor/mongo" {
+
         var Collection: CollectionStatic;
         interface CollectionStatic {
-            new <T>(name: string | null, options?: {
+            new <T, U = T>(name: string | null, options?: {
                 connection?: Object | null;
                 idGeneration?: string;
-                transform?: Function | null;
-            }): Collection<T>;
-        }
-        interface Collection<T> {
-            allow(options: {
-                insert?: (userId: string, doc: T) => boolean;
-                update?: (userId: string, doc: T, fieldNames: string[], modifier: any) => boolean;
-                remove?: (userId: string, doc: T) => boolean;
+                transform?: (doc: T) => U;
+            }): Collection<T, U>;
+        }
+        interface Collection<T, U = T> {
+            allow<Fn extends TransformFn<T> | undefined = undefined, Doc = DispatchTransform<Fn, T, U>>(options: {
+                insert?: (userId: string, doc: Doc) => boolean;
+                update?: (userId: string, doc: Doc, fieldNames: string[], modifier: any) => boolean;
+                remove?: (userId: string, doc: Doc) => boolean;
                 fetch?: string[];
-                transform?: Function | null;
+                transform?: Fn;
             }): boolean;
-            deny(options: {
-                insert?: (userId: string, doc: T) => boolean;
-                update?: (userId: string, doc: T, fieldNames: string[], modifier: any) => boolean;
-                remove?: (userId: string, doc: T) => boolean;
+            deny<Fn extends TransformFn<T> | undefined = undefined, Doc = DispatchTransform<Fn, T, U>>(options: {
+                insert?: (userId: string, doc: Doc) => boolean;
+                update?: (userId: string, doc: Doc, fieldNames: string[], modifier: any) => boolean;
+                remove?: (userId: string, doc: Doc) => boolean;
                 fetch?: string[];
-                transform?: Function | null;
+                transform?: Fn;
             }): boolean;
-            find(selector?: Selector<T> | ObjectID | string, options?: {
-                sort?: SortSpecifier;
-                skip?: number;
-                limit?: number;
-                fields?: FieldSpecifier;
-                reactive?: boolean;
-                transform?: Function | null;
-            }): Cursor<T>;
-            findOne(selector?: Selector<T> | ObjectID | string, options?: {
-                sort?: SortSpecifier;
-                skip?: number;
-                fields?: FieldSpecifier;
-                reactive?: boolean;
-                transform?: Function | null;
-            }): T | undefined;
+            find(selector?: Selector<T> | ObjectID | string): Cursor<T, U>;
+            find<O extends Options<T>>(selector?: Selector<T> | ObjectID | string, options?: O): Cursor<T, DispatchTransform<O['transform'], T, U>>;
```

Comment:
this is very hard to read and understand, and to maintain in the future
I would suggest to simplify it
maybe like this? ⬇️ 

```ts
type FindOptions<T, U> = {
    sort?: SortSpecifier;
    skip?: number;
    limit?: number;
    fields?: FieldSpecifier;
    reactive?: boolean;
    transform?: (doc: T) => U;
}
type FindOneOptions<T, U> = Omit<FindOptions<T, U>, "limit">;

find<Y = U>(selector?: Selector<T> | ObjectID | string, options?: FindOptions<T, Y>): Cursor<T, Y>
findOne<Y = U>(selector?: Selector<T> | ObjectID | string, options?: FindOneOptions<T, Y>): Cursor<T, Y>
```

I [tested](https://www.typescriptlang.org/play?#code/C4TwDgpgBAyg9gJ2DSBjAlgM3RBUC8UAhgHYgBQokUAYjgDYAmKEG2uBxZl40ME9VsEQAeACoA+TqQpVoAeQBGAKyEBJACLTucqAGEArggDOosQBooAVSmEA3lEwRgqABYAuKAAoAlASlWANoAulAAvuSRunQkjPJgwOhwJMbiljacduRQOVCmSAD8nvBILGw4CADc2bnGANboYEVQJAYAtoq41blQ9Oht6MDNrR1dNTnsTMbNdALMaFgV3bkIEESoiQBuEM2KcHCCpMs5wAikxpiIbc1ejHConmJ++AHVEdHosfIkEPGJyakLNZbFB5ANgCIYnEEkkUmlgZYAER9cGIiTVcifYC4TDraB6A6CDaw+FWTiSKBZHrYWIiACanBsXmMAiEiGa-CJwgQ4ikAB9QSp1FoBcZTp8AOaWOAwgEzT7Q-5woF0iQ+TyGExmSyq8aOBXfCD0xkSZmsjbs4rm7m8qACpSqDaaO15cUkKVQGVK6aeKGGv6wwE6tUaoz5eG6iI8ahWFl4ez6kzAAD6JCIbQgnjFCEllnoRDFqfTmddOfdlhZAA9PIi2kRBIiXYinHWG+FqrpY7gAKKVxlxqAAMkpjgM9HoRYzWbdEvb0egXYQvYATP2OMOHG0EMnEMmt8Zp2XZ2FqoJgNY4wf9IShCTF+k472QXYwsRjFwQBjMAYSMTklAAFkiDARcxDgRde1ue5PEXPwqRWZwjBIEcADo0LuVBLG-cdJxLDCUOwJNcKgABqKBG0bMj8PzQs0wzOcIm-X8lUA4DQPAx9K2XKCHgvXA4L1VZgCQ1D0PuSwtx3bd908fCqwIfBCFresIEbApyIAhBGxrTTjEbE9yCjVAAXPJwXFcCBGDXEwCIVLwX0sBxTnOS4EDaTwgJAuMwIgrjwh8AjnDcXwMWMlIDggFD6DgCUvDMtxLMCAAGYIUMk3d9x8aogA) and it seems TS is able to infer `Y` based on the transform function, so no need to explicitly define it

Reply:
As above we should check to null